### DISCUSSION TITLE
QUIC 服务器开加密 客户端不开加密 限制域名 竟然能连接

### DISCUSSION THREAD

**Discussion by k79e** [v2fly/v2ray-core#839] [Q&A] [answered]
v4.36.2
我服务器配置的一个证书 肯定限制域名的 不是自动生成的
客户端填错域名信息 报错无法连接 域名不对

改对了能连接

然后奇怪的问题出现了
客户端关闭tls 竟然也能连接. 这是为啥?

乃至改错域名 无法连接 然后关了tls或者注释掉他 又立刻能连接了.
可是我tls设置里面是不允许非安全的 "allowInsecure": false,
我只注释了security 而tls设置没管它

按理说tls关了 tls设置也应该不生效. 那么既然没域名信息 他又是什么连接上的? (我连接地址填的ip)
而且就算生效 我域名写的就是错的 这时候他怎么能接上!??
恐怕是allowinsecure出了什么问题?

是不是客户端关闭tls 然后开启tls的时候 自动还会走加密
然而这时候 tls设置被忽略了 (默认是允许不安全连接的) 所以就连接上了 是这样么?

是他自己获取服务器证书然后使用了对应的域名么? 看起来只有这个可能性了

**Comment by bhoppi** [comment]
两句话总结QUIC与TLS配置的关系：
当TLS配置合法有效并启用时，QUIC就读取并使用你的TLS设置。这时如果域名对不上，又有`"allowInsecure": false`的设定（默认），就会报错无法连接。
否则，当TLS配置不合法或者未启用时，QUIC就忽略掉你的配置，内部生成一个域名为`quic.internal.example.com`的证书，并设置`"allowInsecure": true`。此时不对域名进行验证，可以连接上，但并不安全。

**Comment by k79e** [comment]
配置很合法
测试是修改的客户端配置 不是服务端

**Reply by bhoppi** [reply]
我说的是一般性原则，不单单针对你的情况。你没有非法配置的情况，但有不启用tls的情况，对号入座即可。
另外，在v2ray的TLS配置里，`allowInsecure`就是只对客户端有效，服务端从不验证客户端的证书因此服务端配置了也没意义。

**Reply by k79e** [reply]
没有不启用
服务器开启tls的 仔细看下描述. 
我是客户端关tls 服务器开tls 而且必须域名正确才可以连接
1楼不是都提了么 双边tls 客户端错域名 连不上 证明配置没问题
然后客户端关了tls 但是还能连接

**Reply by bhoppi** [reply]
我说的不启用TLS具体指“客户端的配置中不启用TLS”。
看来我讲一般性原则不能让你明白，那我就针对你的情况具体说一下吧：
1. 在客户端配置中开启TLS，并进行正确的配置，这时可以连上；如果填错了域名信息，由于有`"allowInsecure": false`的限制，无法连接。这两点都很好理解。
2. 在客户端配置中关闭TLS，虽然tlsSettings保持正确设置，但是这些配置项会被忽略。当使用QUIC协议发起连接时，由于QUIC要求必须启用TLS，因此v2ray内部又自己生成TLS配置，其中假定ServerName为"quic.internal.example.com"，同时设定`"allowInsecure": true`。这样，TLS协商过程中客户端不对ServerName做验证，因此ServerName虽然对不上号，但仍然能够连接，只不过这个连接并不安全。

所以并不是你说的“自己获取服务器证书然后使用了对应的域名”，仅仅就是内部设定`"allowInsecure": true`然后发起不安全连接，仅此而已。

**Reply by k79e** [reply]
客户端生成证书有意义么??
证书不是服务器用的么

**Reply by bhoppi** [reply]
是的，客户端不会生成证书。你可以忽略我前面说的生成证书这句话，这是针对一般情况说的，是指如果服务端没有合法的TLS配置，则服务端会为QUIC协议自动生成CN为`quic.internal.example.com`的自签名证书。你这里不涉及到这个操作。

**Reply by k79e** [reply]
你直接说客户端关了tls也关不掉 会自动开启tls  走默认配置不就完事了...
然后默认配置就是不校验证书的..

**Reply by bhoppi** [reply]
这么理解是不对的。默认的TLS配置allowInsecure的值是false，也就是不允许非安全连接。
当且仅当使用QUIC协议并且读取不到TLS配置信息时，自动生成的TLS配置内特别包含了`"allowInsecure": true`这一条。
使用其他传输协议时，就不会存在这个问题。
所以楼下 @database64128 才会这么提议。

**Reply by k79e** [reply]
我还以为默认是不校验呢 天哪 这果然是bug!!!! 他为啥不按默认的走呢!!!

**Reply by bhoppi** [reply]
这的确是个安全隐患。
我猜测原作者是为了让QUIC协议能够“开箱即用”，所以添加了自动生成TLS配置的功能，但自动生成的证书显然不能通过校验，所以又加了`"allowInsecure": true`。

**Reply by k79e** [reply]
真是方便的有些吓人了 :D

**Comment by database64128** [comment, marked as answer]
The current behavior when the client doesn't have TLS configured is dangerous. I think we should change it to let v2ray panic at startup when using QUIC without TLS configured.

**Reply by k79e** [reply]
It will automatic enable tls but won't verify certificate. (on client)

**Reply by k79e** [reply]
Maybe should show a warn that default tls setting is used to prevent accident wrong configurate.
well I'm wrong the default setting will verify certificate!!!
